src/clusterfuzz/_internal/bot/tasks/trials.py

```diff
@@ -61,17 +75,25 @@ def __init__(self):
       for config in trials_config:
         if config['app_name'] != app_name:
           continue
-        self.trials[config['app_args']] = config['probability']
+        self.trials[config['app_args']] = AppArgs(config['probability'],
+                                                  config.get('contradicts', []))
     except Exception as e:
       logs.log_warn('Unable to parse config file: %s' % str(e))
       return
 
   def setup_additional_args_for_app(self):
     """Select additional args for the specified app at random."""
-    trial_args = [
-        app_args for app_args, probability in self.trials.items()
-        if random.random() < probability
-    ]
+    trial_args = []
+    contradicts = set()
+
+    for app_args, flag_data in sorted(
+        self.trials.items(), key=lambda x: random.random()):
+      if (random.random() < flag_data.probability and
+          app_args not in contradicts and
+          not any(flag in flag_data.contradicts for flag in trial_args)):
```

Comment:
Not sure if I'm maybe micro-optimizing here, but this check makes the algorithm quadratic (for cases where above expressions are true). I'm not sure if Python will build an optimization for the case where flag_data.contradicts is empty.

A simple guard for this would be:
(not flag_data.contradicts or the rest of your line)

Since this gets a bit convoluted, maybe extract it into a function?

Reply:
done